File: src/TurboOmicsIntegrator/scripts/py/pca_anova_analysis.py
import argparse
import os
import sys
import pandas as pd
from sklearn.decomposition import PCA
import json
import logging
from anova_analysis import get_anova_tukey

#
# Constants
#

MAX_COMPONENTS = 10


#
# Local Functions
#

def create_folder_if_not_exists(folder_path):
    if not os.path.exists(folder_path):
        try:
            os.makedirs(folder_path)
            print(f"Folder '{folder_path}' created successfully.")
        except OSError as e:
            print(f"Error creating folder '{folder_path}': {e}")
    else:
        print(f"Folder '{folder_path}' already exists.")


# Read the JSON formatted DataFrame
def read_dataframe(file_name):
    try:
        df = pd.read_json(file_name, orient='records')
        return df
    except FileNotFoundError:
        logging.error(f"The file '{file_name}' does not exist.")
        return None

# Perform PCA and save the results in JSON files
def perform_pca_and_save(x, outfolder_path):
    # Create a PCA object with the desired number of components
    n_components = min([MAX_COMPONENTS, x.shape[0], x.shape[1]])
    pca = PCA(n_components=n_components)
    
    # Perform PCA on the DataFrame's data
    principal_components = pca.fit_transform(x)
    
    # Create a DataFrame for the projections
    projections_df = pd.DataFrame(principal_components, columns=[i+1 for i in range(n_components)], index=x.index)
    
    # Create a DataFrame for the loadings
    loadings_df = pd.DataFrame(pca.components_.T, columns=[i+1 for i in range(n_components)], index=x.columns)
    
    # Calculate the fraction of variability explained by each component
    explained_variance = pca.explained_variance_ratio_
    explained_variance_df = pd.DataFrame({'Explained_Variance': explained_variance}, index=projections_df.columns)
        
    return projections_df, loadings_df, explained_variance_df

#
# Main
#

def main(args):
    
    omic, x_filepath, mdata_filepath, mdataType_filepath, index_filepath, outfolder_path =\
        args.omic, args.x_filepath, args.mdata_filepath, args.mdataType_filepath, args.index_filepath, args.outfolder_path
    
    #
    # Read tables
    #
    logging.info('Reading tables...')
    
    with open(index_filepath, 'r') as f:
        myIndex = json.load(f)
    logging.info(f'index file read: {index_filepath}')
        
    with open(mdataType_filepath, 'r') as f:
        mdataType = json.load(f)
    logging.info(f'mdataType file read: {mdataType_filepath}')
    
    x = read_dataframe(x_filepath)
    logging.info(f'x file read: {x_filepath} read')
    
    mdata = read_dataframe(mdata_filepath)
    logging.info(f'mdataType file read: {mdataType_filepath}')
    

    #
    # Set indexes
    #
    x.index = myIndex[f'x{omic}']
    mdata.index = myIndex['mdata']
    
    
    #
    # Calculate PCA
    #
    logging.info('Calculating PCA...')
    projections_df, loadings_df, explained_variance_df = perform_pca_and_save(
        x, 
        outfolder_path
        )
    logging.info("PCA completed")

    # 
    # Writing PCA output
    #
    logging.info('Writing PCA output...')
    projections_df.to_json(os.path.join(outfolder_path, "projections.json"), orient='index')
    loadings_df.to_json(os.path.join(outfolder_path, "loadings.json"), orient='index')
    explained_variance_df.to_json(os.path.join(outfolder_path, "explained_variance.json"), orient='index')
    
    #
    # Apply ANOVA
    #
    logging.info('Calculating ANOVA and Tukey-HSD...')
    anova_res = get_anova_tukey(projections_df, mdata, mdataType, outfolder_path)
    with open(os.path.join(outfolder_path, 'anova.json'), 'w') as f:
        json.dump(anova_res, f)
    logging.info("ANOVA and Tukey-HSD completed")
    
    return 0


if __name__ == "__main__":

    parser = argparse.ArgumentParser(description="""
        PCA, ANOVA and Tukey-HSD analysis. 
        Exec example: 
            python pca_anova_analysis.py omicType path/to/x path/to/mdata path/to/mdataType path/to/index path/to/outfolder
    """)
    
    parser.add_argument("omic", help="Omic type {q, m}")
    parser.add_argument("x_filepath", help="Path to file with quantitative data")
    parser.add_argument("mdata_filepath", help="Path to file with mdata")
    parser.add_argument("mdataType_filepath", help="Path to file with mdataType")
    parser.add_argument("index_filepath", help="Path to file with index")
    parser.add_argument("outfolder_path", help="Path to output folder where files will be written")
    args = parser.parse_args()
    
    create_folder_if_not_exists(args.outfolder_path)
    
    logFile=f'{args.outfolder_path}/pca_anova_analysis.log'
    print(logFile)
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(levelname)s - %(message)s',
        handlers=[
            logging.StreamHandler(stream=sys.stdout),
            #logging.FileHandler(logFile)
        ]
    )
    
    logging.info('Start main')
    main(args)
    logging.info('End main')

File: src/TurboOmicsIntegrator/scripts/js/pcaAnovaAnalysis.js
/*
Import modules
*/
const path = require('path')
const { spawn } = require("child_process");
const fs = require("fs");

/*
Constants
*/

// Files that must be generated after the execution
const pcaFiles = [
    'anova.json',
    'explained_variance.json',
    'loadings.json',
    'projections.json'
]

/*
Local function
*/
function checkFileExistence(myPath, files) {
    for (const file of files) {
        console.log(file)
        if (!fs.existsSync(path.join(myPath, file))) {
            return false;
        }
    }
    return true;
}

/*
Main function
*/
function PCA_ANOVA_PY(myPathX, myPathPCA, omic, myLogging) {

    return new Promise(resolve => {

        myLogging(`Executing PCA_ANOVA_PY in ${omic} omic`);
        const script = 'pca_anova_analysis.py'

        fs.writeFile(
            path.join(myPathPCA, omic, '.status'),
            JSON.stringify({ status: 'waiting' }),
            () => resolve(0)
        );

        const process = spawn(
            global.pythonPath,
            [
                path.join(__dirname, `../../scripts/py/${script}`),
                omic,
                path.join(myPathX, `x${omic}_norm.json`),
                path.join(myPathX, `mdata.json`),
                path.join(myPathX, `mdataType.json`),
                path.join(myPathX, `index.json`),
                path.join(myPathPCA, omic)
            ],
            { encoding: 'utf-8' }
        );

        process.stdout.on('data', data => fs.appendFile(
            path.join(myPathPCA, omic, '.log'),
            `stdout: ${data}`, () => { })
        );

        process.stderr.on('data', data => fs.appendFile(
            path.join(myPathPCA, omic, '.log'),
            `stderr: ${data}`, () => { })
        );

        process.on('close', code => {
            myLogging(`${script} process in ${omic} omic exited with code ${code}`);
            if (
                code == 0 &&
                checkFileExistence(
                    path.join(myPathPCA, omic),
                    pcaFiles
                )
            ) {
                fs.writeFileSync(
                    path.join(myPathPCA, omic, '.status'),
                    JSON.stringify({ status: 'ok', code: code })
                )
            }
            else {
                fs.writeFileSync(
                    path.join(myPathPCA, omic, '.status'),
                    JSON.stringify({ status: 'error', code: code })
                )
            }
        })
    })
}

module.exports = PCA_ANOVA_PY;